### ISSUE TITLE
Autodoc software modules are needed

### ISSUE THREAD

**Issue by Firestorm01X2** [CleverRaven/Cataclysm-DDA#27750]
**Describe the bug**  
Old chances to install or uninstall CBMs were designed around idea of player able to install CBMs with hands. But now it is changed.

Currently player uses special device aks Autodoc. But at the same time skill requirment basically remain the same.

In addtion player has to bring anastetic kits.
 
Over all  this shifts makes CBM instalation much harder than it was before. 

Also fact of using Autodoc does not reflect on install chances.

Something has to be done.

**To Reproduce**  
Try to install simplest CBM like Power Storage CBM

**Expected behavior**  
Player able to find some kind of way that automates process of installation of such simple CBM without chance to failure. Or with almost zero chance to failure.

**Actual behavior**  
Player always have to setup installation manually with chances to failure even for simpliest CBMs. With low skills player basically unable to install CBMs without huge risk.

**Ways to solve**  

**1 ) Autodoc software modules**  
https://github.com/CleverRaven/Cataclysm-DDA/pull/22877#issuecomment-362961224
> kevingranade:
>Player acquires a program for installing a specific CBM (i.e. on a USB stick).
>This would have an almost 0 chance of bad outcomes.
>This also gives low-int characters a chance of installing, the requirements to just load the software should be very easy.
>Players would find these in lab databases, loaded into autodocs, or perhaps rarely elsewhere on USB sticks.
>This might have qualifications for when it will work, such as not working if the targeted body part is mutated. It will still be safe, but it will refuse to work.

All CBMs have difficulty associeted with it.
Idea of creating software USB sticks that contains software for fully automated CMB installation of certain difficulty.

For example let's introduce 3 types of USB sticks

1) For installation of  civillian CBMS (difficulty 4 and below). Should be very common.
2) For installation of heavy duty CMBS (difficulty 8 and below.). Should be found in medical locations.
2) For installation of military CMBS (difficulty 15 and below.) Should be found in military locations or labs.

If player has this stick into into inventory (or somewhere around) then CMB installation of supported difficulty will be 100% guaranteed.

Likely slight rebalance of CBM diffulty stats are required.

I have to note that is has not to be full on new Autodoc programming system. Software USB sticks can just checked for presence like anesthetic kits.

Basic idea where to start can be found here:
https://github.com/CleverRaven/Cataclysm-DDA/pull/27623

PROS
+Explainable solution.
CONS
-Adding new system to solve balance problem.

**2 ) Increase chances of CBM installation using autodocs**  
Change installation chances to reflect using of autodoc. We are pretending that software already there and simplify installation a lot.
PROS
+Simple soulution
CONS
-Probably too straightforward.

**3 ) Making installation of CMBs using autodoc 100% success**  
We are pretending that software already make installation process safe.
PROS
+Simpliest solution
CONS
-Completely neglects CBM installation failures and failure bionic system.
-Probably too straightforward.

**Comment by Night-Pryanik** [comment]
> Also fact of using Autodoc does not reflect on install chances.

I don't get it. Care to explain?

> With low skills player basically unable to install CBMs without huge risk.

As I said here (https://www.reddit.com/r/cataclysmdda/comments/aet3jp/autodocs_literally_make_no_sense_gameplay_and/edvpl08/), Power Storage CBM could be installed with a fairly high chance of success even by a complete noob with 2 in first aid, computers and electronics and default 8 Intelligence.

> If player has this stick into into inventory (or aroun player) than CMB installation of supported difficulty will be 100% guaranteed.

I'm against this. At least for the proposed three grades of software modules. That will make bionic installation MUCH easier, even compared to the old manual-installation system. Finding military-grade software module would almost mean "won the game", as the only requirement for bionic installation would be only anesthetics.

I'm in much more favor of "per-bionic" software module system. This way one can achieve 100% chance of success of installation only for the CBM for which this module was designed. Thus finding "Power Storage CBM software module", for example, won't affect chances of success for installation of Time Dilation CBM.

**Comment by Firestorm01X2** [comment]
>I don't get it. Care to explain?

Installation chances was balanced around the fact of restricting CBM installation with skills. It is from old pre Autodoc changes. Currently there is additional restrictions but chacnes of failure are old.

**Comment by Night-Pryanik** [comment]
Possible explanation: higher-tier software modules are enhanced diagnosis (more info could be read from sensors, more reliable data concerning the patient could be built from this info) and prediction (better choice of CBM placement in the body) programs, superior to the ones installed in basic autodoc, that improve overall autodoc performance by reducing the chance to fail the installation.

Maybe we don't need to add the software modules, but instead the hardware modules? Upgrading the basic autodoc with some better modules with surely reduce the chance to fail.

**Comment by nexusmrsep** [comment]
If a CBM is a self contained all-in-one packet, as it is said to be, then why is it not equipped with a relevant software in the first place? My guess here would be: it is. Thus either software packs would have to be some high grade military tech unknown to general public for obvious reasons, or some bootleg hack tools of some home-brew cyberpunk hacker-medics, that try to invent something better on the fly. The second option would suggest that you may also find bonkers soft, that is worse then original, or even a trap software that will mess you up.

Generally I don't like the idea of software for CBM's. The logic doesn't click to me. As I said in a different thread, rather then coding the machine, I believe the whole autodoc setup is rather a human operated pre-surgery diagnostic and adjustment process. Imagine same derma-armor installed on a man versus same installation on a woman - it needs personal settings to be done first. The progression of difficulty is obvious as it comes to more and more invasive operations, with bionic systems that either augment or replace vital senses, organs, or in general create a new quality that has to be linked and synced with a living organism. Consider the difference between following surgeries: removing an ingrowing toenail, appendix removal in appendicitis, and on the high end: cardio & neurosurgery. If those operations require different levels of skill and preparations, than there is no reason why there should there not be a difference between installing a Power Storage CBM and a Cerebral Booster CBM, etc. No software would help in this.

Instead, I think bionic slots should be implemented ASAP. It's a ready PR, that should be put in its place (merged) now instead of waiting for the illusory milestone of 0.D
With them in place the whole CBM system would achieve a new balance, that might be then tweaked by perhaps loosening other constrains that were placed for balance only.

**Comment by Firestorm01X2** [comment]
>This is already the case, the NPC in the farm settlement will install CBMs for you.

NPCs are disabled by default. New player wouldn't even seen them without tweaking default settings.

**Comment by Night-Pryanik** [comment]
> If a CBM is a self contained all-in-one packet, as it is said to be

It is not. CBM as all-in-one kit is purely fanon, as it wasn't ever described as such. CBM kit contain just bionic designed to be installed only by a skilled doctor (or on a pre-configured autodoc), and little of else.

**Comment by Night-Pryanik** [comment]
> I feel like if it needs new hardware modules, they'd mostly be hard blockers on installation, not just a bump to install chance.

I like the idea more and more. This way we could even allow the long-requested feature of dismantling an autodoc in some distant location and assemble it in some comfortable place.

**Comment by Firestorm01X2** [comment]
@nexusmrsep 
>Instead, I think bionic slots should be implemented ASAP.

Ok. But what to do next?

**Comment by Firestorm01X2** [comment]
>This is a hard no. You even quoted me, but missed the point, I said "Player acquires a program for installing a specific CBM (i.e. on a USB stick)." A SPECIFIC CBM. Not whole categories of CBMs such that they need to just find three widgets and then they're done forever.
>I feel like if it needs new hardware modules, they'd mostly be hard blockers on installation, not just a bump to install chance.

Ok. It may works.

But:
Occam's razor
https://en.wikipedia.org/wiki/Occam%27s_razor
>Occam's razor is the problem-solving principle that essentially states that simpler solutions are more likely to be correct than complex ones.

If we pretend that 
1) CBMs already contains software
2) CBMs should be twice as rare
3) Chance to install should be higher

then result rougly will be the same.

Probably such sofware/hardware modules should support at least 3 CBMs or it will be just overkill (look Occam's razor).

**Comment by Firestorm01X2** [comment]
@Night-Pryanik 
@kevingranade 
@nexusmrsep 

Alternate idea:
1) Enable CBM slots
2) Make installation much easier
3) Make some really powerfull CBMs harder to find.

This way will tempted to constantly switch CBMs, adapting to the current situation.

**Comment by Amoebka** [comment]
3 USB sticks is too little, one per CBM is too many, in my humble opinion.
What about adding USBs based on already existing bionic professions? For example, an "Industrial Cyborg Software Pack" can include installation guides (100% success with autodoc, still need bionic itself and anesthetic) for all CBMs Industrial Cyborg starts with, "Bionic Firefighter Pack" - all CBMs Bionic Firefighter starts with, etc. Would make a lot of sense too, after all these professions appear to be fairly common in 2050, and their bionic outfits would be standartized. Natually, stuff like Bionic Assassin or Broken Cyborg won't have USB sticks. There will be quite a lot of overlap, but I think that's fine? Bionics that aren't found in any profession can be covered by their own unique USBs (especially something experimental like Time Dilation).

**Comment by SirPendrak** [comment]
Definitely yes for CBM slots, but bring it with a mod that disables limits for peps who like no limits.  
Grouped pendrives for some low lvl CBMs and single ones for advanced seems reasonable, could help new characters, and become another trophy hunt for late characters. Might be a little too RNG to actually affect gameplay much, but its better than nothing. And there should be a way to download software to E-ink tablet. And maybe CBM software should not only be found on pendrives, but also sd-cards and computers in hospitals/labs/bunkers etc. Maybe rarerly in other mundane places. People tend to have strange things in their personal data sometimes.

**Comment by Night-Pryanik** [comment]
> Do autodocs have MRI scanners built in?

Sort of. Some scanners are built-in for sure.

**Comment by Firestorm01X2** [comment]
Let's start with this:
https://github.com/CleverRaven/Cataclysm-DDA/pull/27780

It can be used as balance tool. Now we can easier CBM installation and player still will be thinking what to install and when to install it.

**Comment by Night-Pryanik** [comment]
> Can I soften CBM installation skill requrments there?

No. It should be done in a separate PR.

Furthermore, please do tell why do you think there is a need for softer requirements. With usual "expected vs. actual" comparison, numbers, calculations etc.

**Comment by Firestorm01X2** [comment]
>Furthermore, please do tell why do you think there is a need for softer requirements. With usual "expected vs. actual" comparison, numbers, calculations etc.

To balance new restriction. Guranteed way to install CBMS are big TODO for now.

With slot system player should constantly tweak himself to fit current needs. It will be hard until raising skills.

**Comment by nexusmrsep** [comment]
> Ok. But what to do next?
> Alternate idea: (...)

There are few competing concepts with CBM's high risk - high reward balance:
- benefits
- rarity
- risks
- realism

As we all know benefits from CBMs are huge. Installing a few is either the endgame itself, as in "no more incentive", as you get near-invincible, or at least game changing - you change the play-style to utilize the maximum from what you have installed.

This is offset by the rarity - which is not as terrible as it seems as long as you keep a scalpel and raid a few labs. The rarity of the anesthetic kits is far more limiting. Proof: at least few of latest Vormithrax runs shows that you can get more CBMs that you can install. It's not a bad concept in itself for balancing reasons, and it absolutely satisfies the realism factor, but personally I'd rather see bionic slots as a limiting factor here.  Why? Because with bionic slots you'd need to specialize, so potential abundance of CBM's would not thwart balance, but would be even somewhat desired (in  reasonable boundaries). You wouldn't be able to install them all anyway. This in fact may even reduce the need for anesthesia kits to be so rare.

As for risks, I believe that they create lot of fun in this game. Going towards "just find CBM and you're in the clear", is in my opinion watering down the challenge. There is no controversy in mutations being risky per RNG factor, so I seriously don't understand what is so controversial in valuable CBMs in being risky per installation? For one this is great as per realism factor, for the medical reasons I stated before. Secondly it encourages exploration, preparation and taking other risks, with a great reward at hand. Third thing: failure can itself become a game challenge. This drives the game-play forward. I do understand that it might not be seen as desired from a player's perspective, but its realistic in a sense that you don't control everything that happens in life, and are often forced off the desired path to circumvent the problems. That said, perhaps it would be beneficial to think of reducing the requirements for sorting out the negative effects of bad installations, so risk won't be so severe or at least reversible with effort. Like for example introducing "ordinary" CBM prosthetic implants of limbs, eyes, etc. Lost your arm due to installation? Replace it with a robotic limb. Broken leg? Robotic leg never breaks, but needs to repaired when damaged. (Seems that is an interesting idea on it's own.)
 
As for realism (setting aside the CBMs themselves as they fall under a self-explanatory S-F theme):

- Self installation is a no go for me. Using autodoc and NPC's is absolutely real. You'd barely be able to sew your own lacerated skin, so precise operation of own organs are absurd. 
- Bionic slot system is a must. Human body is not a balloon to stick as much inside as you possibly can.
- Anesthesia fits well with realism, but might have become a choking collar because of rarity

And finally the problem of skills used in installation. Or rather problems. One is that first-aid skills is one of the depreciated and less versatile skills. Justice calls for it being applied somewhere, and medical aspects of CBM installations are a great place where it can shine. Its a bit less true for electronics/computers since they lot more areas of effect. Second is that those skills are used in determining success of installation. That is not bad if you think about the fact that the risk itself is desired, and that binary "0 or 1" , "true or false" model is bad. And that also pushes for actions - player wants to hone those skills to improve chances, and that drives game-play. 

That is the reason why the complete removal of the skill content in favor of finding dedicated software for installation is not appealing to me.. It might perhaps be good as as suplement - so you either use skills to program the autodoc **OR** use CBM-specific soft to program the autodoc.

This however is a risk in itself. CBMs would become a very complicated system. You'd need too many things to "click" together at once (find CBM -> find autodoc -> find anesthesia -> find soft ). If RNG fails at each stage - you are screwed, and I didn't even mention slots at this point. 

In conclusion I'd see it as following:

1. Enable CBM slots.
2. Keep current CBM availability. 
3. Make anesthesia easier to get (5 per stack instead of 3 perhaps?) as slots balance the top end of the spectrum.
4. Keep skill requirements for the installation. Use USB & SD cards medical soft as **alternative** for skill requirement, so you use either option. Use CBM-specific soft only for that purpose. Double it's function as a blueprint for nano-fabricator for CBM production.

**Comment by Firestorm01X2** [comment]
>Make anesthesia easier to get (5 per stack instead of 3 perhaps?) as slots balance the top end of the spectrum.

This is make sense now.

**Comment by Firestorm01X2** [comment]
>Enable CBM slots.
https://github.com/CleverRaven/Cataclysm-DDA/pull/27780

>Make anesthesia easier to get (5 per stack instead of 3 perhaps?) as slots balance the top end of the spectrum.
https://github.com/CleverRaven/Cataclysm-DDA/pull/27784

**Comment by kevingranade** [comment]
There is no alternate idea, you need to do this:
> You need to outline the current failure rates and what you expect those failure rates to be, because it's unclear if there is a problem and the extent to which there is a problem.

Before you do anything else in this area.

**Comment by Firestorm01X2** [comment]
@kevingranade 
>You need to outline the current failure rates and what you expect those failure rates to be, because it's unclear if there is a problem and the extent to which there is a problem.
>Before you do anything else in this area.

Maybe it is not even needed anymore. 
Look at the @nexusmrsep  answer: https://github.com/CleverRaven/Cataclysm-DDA/issues/27750#issuecomment-456498648

His ideas looks pretty solid.
There is only approval needed.

**Comment by Firestorm01X2** [comment]
>Keep skill requirements for the installation. Use USB & SD cards medical soft as alternative for skill requirement, so you use either option. Use CBM-specific soft only for that purpose. Double it's function as a blueprint for nano-fabricator for CBM production.

So I imagine thic USB drives, like:

```
{
    "abstract": "bionic_software_general",
    "type": "BIONIC_SOFTWARE",
    "category": "bionics",
    "name": "abstract bionic software module",
    "weight": 2041,
    "volume": 10,
    "bashing": 8,
    "material": [ "steel", "plastic" ],
    "symbol": ":",
    "color": "yellow",
    "use_action": "install_bionic",
    "flags": [ "CBM_SOFTWARE" ]
  },

{
    "id": "bio_power_storage_software",
    "copy-from": "bionic_software_general",
    "type": "BIONIC_SOFTWARE",
    "name": "Bionic Power Source software",
    "description": "This USB drive contains pre-trained neuronal network data for safe installing of various power banks",
    "price": 4000,
   "bionic_ids": [ "bio_power_storage", ""bio_power_storage_mkII"" ]
  },

{
    "id": "bio_targeting_software",
    "copy-from": "bionic_software_general",
    "type": "BIONIC_SOFTWARE",
    "name": "Targeting System CBM software",
    "description": "This USB drive contains pre-trained neuronal network data for installing hight quality bionic optics",
    "price": 4000,
   "bionic_ids": [ "bio_targeting" ]
  },

{
    "id": "bio_leaky_software",
    "copy-from": "bionic_software_general",
    "type": "BIONIC_SOFTWARE",
    "name": "Emergency leaky bionic removal software",
    "description": "This USB drive contains pre-trained neuronal network data for emergency removal of leaky bionics.",
    "price": 4000,
   "bionic_ids": [ "bio_leaky" ]
  },
```

**Comment by Fris0uman** [comment]
>Use USB & SD cards medical soft as alternative for skill requirement, so you use either option

Why would pre-programmed surgery on usb even exist in the first place ? I mean aside from any balance concern. The auto-docs are machines operated by trained professional, if pre-programmed settings exists they are already in the machine not laying around on flash drives.
Unless they are some weird atempt at making basement surgery more successfull ...
 That said, the idea of bootleg auto-doc settings that might help you OR mess you up even more seems pretty fun, it would not affect balance too much as there is still RNG in it.
It could work like mushroom used to : they all look the same but if you have enought skill you can identify the bad ones.
It doesn't solve anything regarding CBM balance and all but it makes more sense as why those usb exists.

**Comment by Firestorm01X2** [comment]
>Why would pre-programmed surgery on usb even exist in the first place ? 

It can. Obviously, 
But I personally have no idea on how to implement that. Aside from simple spawning USB drives near some Autodocs.

**Comment by Fris0uman** [comment]
I mostly wanted to point out that, IMO, the idea of auto-doc sofware modules in itself doesn't make much sense. Because I'm pretty sure that pre-made settings on a usb stick would not be a thing for something as high rech as the autodoc.
I can't even find pre made settings for a 3D printer irl, never mind a robotic surgery assistant.

**Comment by kevingranade** [comment]
> Maybe it is not even needed anymore.

Yes, it is still needed, I'm closing this issue until you provide that information.

> Why would pre-programmed surgery on usb even exist in the first place ?

From the original suggestion:
> Players would find these in lab databases, loaded into autodocs, or perhaps rarely elsewhere on USB sticks.

Having them loaded on USB sticks was the *last* option listed, but it has been selected in this issue as the only option, probably because it is much easier to impliment.

**Comment by Firestorm01X2** [comment]
@kevingranade 

Here is how it is now:

![cbm_now](https://user-images.githubusercontent.com/17512620/51627278-b5d62380-1f52-11e9-86db-790fb9e5b1d0.png)


Here is how approximately it should be:

![cbm_idea](https://user-images.githubusercontent.com/17512620/51627301-c4243f80-1f52-11e9-8c8b-70576d0fc382.png)

**Comment by Night-Pryanik** [comment]
The formula absolutely has to be corrected: absolutely noob character with zero levels in all three required skills gets a whooping 50% chance to successfully configure autodoc to install bionic (even the low-level like Power Storage CBM)? I'd say that's just improbable. And you're saying 

>With low skills player basically unable to install CBMs without huge risk.

?

>But player should not be best Autodoc specialist in the world to reliable install all CBMs.

Right the opposite. Only the best specialists should be able to configure the autodoc so it can reliably install higher-level CBMs.

**Comment by Firestorm01X2** [comment]
If anyone find any mistake in calculation- feel free to show me.

**Comment by Firestorm01X2** [comment]
>Right the opposite. Only the best specialists should be able to configure the autodoc so it can reliably install higher-level CBMs.

Currently this rule is broken. Table indicates that. 70^ in no reliable chance.

**Comment by Firestorm01X2** [comment]
Alternative scaling:
![cbm_idea_2](https://user-images.githubusercontent.com/17512620/51637661-71a44c80-1f6d-11e9-9abe-3568a3211f6c.png)

[CBM_chance.xlsx](https://github.com/CleverRaven/Cataclysm-DDA/files/2789185/CBM_chance.xlsx)

**Comment by kevingranade** [comment]
@Firestorm01X2 my apologies, but you should probably just drop this, because the difference between our expectations is just too large.  I don't see any possible way for us to find a common solution on this.

@Night-Pryanik 
Agreed, no-knowledge chance of successful install is way too high, it needs to be reigned in a great deal, I don't expect anything higher than a 20% success rate there.  At the same time, I do think the requirements at the far upper range are a bit much, I'd expect 10 + 8/8/8 or so to be at least reaching 90% for up to difficulty 10 or so, though I'd buy that the operation to install things like wired reflexes should be absolutely ludicrous and not peak over 90% until you're an absolute virtuoso at everything involved.

A thought I had looking at this is giving the stats some context.  Medical/first aid is an absolute must, and should maybe just tank your chances if you don't have it, regardless of the rest.  Meanwhile computer and electronics can maybe be substituted for with int because there's a certain amount of "you can just figure it out" for programming the machine and setting up the cbm for installation.

So just a sketch, the install success chance is the min of the install chance for the three skills, and they're calculated differently:
1. First aid is dominant, it's basically determining your ability to plan the operation, so no matter how good you are at the others, it presents a cap on your chances of success.
2. Computers is your ability to actually input your planned operation into the system, if it doesn't meet the requirements, maybe you're a bit prone to mistakes and can't optimize the system as well.  The required level is much lower than the nominal First Aid for a procedure, A First Aid 4 procedure might only require Computers 1. It might also contribute by making the manual process of entering the program faster the higher it is.  At low levels int might completely stand in for computer success chance, but not speed.  i.e. if you take your time you can get it done, but you're going to be slow.
3. Electronics is your ability to set up the machine and CBM for installation.  This is almost routine for easy CBMs, but mistakes are costly in chance of risk.  Again, it might only require Electronics 1 for optimal installs until you get to very difficult CBMs.
IDK, I don't think it's *necessary* to change anything about skill handling, but it's nice to have a mental model of what's happening rather than just numbers.

I don't have any intention in following up on this any time soon, and Firestorm is so far off base it's not even worth discussing it, so unless you feel like picking it up I'm just going to leave this closed.

**Comment by Firestorm01X2** [comment]
Still data that I've bring worth separate issue.
